lib/tokens.js

```diff
@@ -408,7 +408,7 @@ export async function resolve({requester, token, levelOfAssurance}) {
           // token resolved for same requester, return pairwise token
           const record = await _getPairwiseToken({internalId, requester});
           const {pairwiseToken: {value: pairwiseToken}} = record;
-          return {pairwiseToken};
+          return {pairwiseToken, internalId};
```

Comment:
The issue here is that we can only return early if the new flag `allowResolvedInvalidatedTokens` (we need to add this) is set to `true` (the default should be `false`). Setting this flag to `true` will allow *already resolved* but subsequently invalidated tokens to be resolved again.

This means that only if the flag is set to `true`, we can return early here. Otherwise, we need to store the value of the token record and allow the invalidation check to happen below. To do that, we need to a new var (`tokenRecord`) to store the pairwise token record in. It should be declared after the comment `// see if token is already resolved` up above:

```js
// see if token is already resolved
let tokenRecord;

// Note: ...
```

Then instead of `const record =` we'd do `tokenRecord = await _getPairwiseToken` and only destructure and return if the flag is set. Then before throwing just below (`Token already used.`) we need to wrap that in a conditional: `if(!tokenRecord)` because the early return doesn't always happen.

The above will make it so we can use the `tokenRecord` if the token has already been resolved *and* it hasn't been invalidated, after we check that below. There's a comment below (`// token is not yet resolved`) that we should add a new conditional above:

```js
if(tokenRecord) {
  const {pairwiseToken: {value: pairwiseToken}} = tokenRecord;
  return {pairwiseToken, internalId};
}
```

That final bit should ensure that we reuse the already-resolved token in the case that it hasn't been invalidated -- whilst still properly ensuring that we run the invalidation check.